src/spatialdata_plot/pl/utils.py

```diff
@@ -1984,6 +2015,156 @@ def set_zero_in_cmap_to_transparent(cmap: Colormap | str, steps: int | None = No
     return ListedColormap(colors)
 
 
+def _get_extent_and_range_for_datashader_canvas(
+    spatial_element: SpatialElement, coordinate_system: str, ax: Axes, fig_params: FigParams
+) -> tuple[Any, Any, list[Any], list[Any], Any]:
+    extent = get_extent(spatial_element, coordinate_system=coordinate_system)
+    x_ext = [min(0, extent["x"][0]), extent["x"][1]]
+    y_ext = [min(0, extent["y"][0]), extent["y"][1]]
+    previous_xlim = ax.get_xlim()
+    previous_ylim = ax.get_ylim()
+    # increase range if sth larger was rendered on the axis before
+    if _mpl_ax_contains_elements(ax):
+        x_ext = [min(x_ext[0], previous_xlim[0]), max(x_ext[1], previous_xlim[1])]
+        y_ext = (
+            [
+                min(y_ext[0], previous_ylim[1]),
+                max(y_ext[1], previous_ylim[0]),
+            ]
+            if ax.yaxis_inverted()
+            else [
+                min(y_ext[0], previous_ylim[0]),
+                max(y_ext[1], previous_ylim[1]),
+            ]
+        )
+
+    # compute canvas size in pixels close to the actual image size to speed up computation
+    plot_width = x_ext[1] - x_ext[0]
+    plot_height = y_ext[1] - y_ext[0]
+    plot_width_px = int(round(fig_params.fig.get_size_inches()[0] * fig_params.fig.dpi))
+    plot_height_px = int(round(fig_params.fig.get_size_inches()[1] * fig_params.fig.dpi))
+    factor: float
+    factor = np.min([plot_width / plot_width_px, plot_height / plot_height_px])
+    plot_width = int(np.round(plot_width / factor))
+    plot_height = int(np.round(plot_height / factor))
+
+    return plot_width, plot_height, x_ext, y_ext, factor
+
+
+def _create_image_from_datashader_result(
+    ds_result: ds.transfer_functions.Image, factor: float, ax: Axes
+) -> tuple[MaskedArray[np.float64, Any], matplotlib.transforms.CompositeGenericTransform]:
+    # create SpatialImage from datashader output to get it back to original size
+    rgba_image_data = ds_result.to_numpy().base
+    rgba_image_data = np.transpose(rgba_image_data, (2, 0, 1))
+    rgba_image = Image2DModel.parse(
+        rgba_image_data,
+        dims=("c", "y", "x"),
+        transformations={"global": Scale([1, factor, factor], ("c", "y", "x"))},
+    )
+
+    # prepare transformation
+    trans = get_transformation(rgba_image, get_all=True)["global"]
+    affine_trans = trans.to_affine_matrix(input_axes=("x", "y"), output_axes=("x", "y"))
+    trans = mtransforms.Affine2D(matrix=affine_trans)
+    trans_data = trans + ax.transData
```

Comment:
I see this kind of construct coming back more also in `render.py`. Perhaps refactor in its own separate function `prepare_transformation`?